pallets/simple-staking/src/tests.rs

```diff
@@ -0,0 +1,110 @@
+use crate::{mock::*, Error, AuthorInfo};
+use frame_support::{assert_noop, assert_ok, traits::{OnInitialize, Currency}};
+use sp_runtime::traits::BadOrigin;
+use pallet_balances::Error as BalancesError;
+
+#[test]
+fn it_should_set_invulnerables() {
+	new_test_ext().execute_with(|| {
+		let new_handlers = vec![1, 2, 3, 4];
+		assert_ok!(SimpleStaking::set_invulnerables(
+			Origin::signed(RootAccount::get()),
+			new_handlers.clone()
+		));
+		assert_noop!(
+			SimpleStaking::set_invulnerables(Origin::signed(1), new_handlers.clone()),
+			BadOrigin
+		);
+		assert_eq!(SimpleStaking::invulnerables(), new_handlers);
+	});
+}
+
+#[test]
+fn set_allowed_author_count() {
+	new_test_ext().execute_with(|| {
+		assert_ok!(SimpleStaking::set_allowed_author_count(Origin::signed(RootAccount::get()), 7));
+		assert_noop!(SimpleStaking::set_allowed_author_count(Origin::signed(1), 8), BadOrigin);
+		assert_eq!(SimpleStaking::allowed_authors(), 7);
+	});
+}
+
+#[test]
+fn set_author_bond() {
+	new_test_ext().execute_with(|| {
+		assert_ok!(SimpleStaking::set_author_bond(Origin::signed(RootAccount::get()), 7));
+		assert_noop!(SimpleStaking::set_author_bond(Origin::signed(1), 8), BadOrigin);
+		assert_eq!(SimpleStaking::author_bond(), 7);
+	});
+}
+
+#[test]
+fn register_as_author() {
+	new_test_ext().execute_with(|| {
+		assert_noop!(
+			SimpleStaking::register_as_author(Origin::signed(1)),
+			Error::<Test>::TooManyAuthors
+		);
+
+		assert_ok!(SimpleStaking::set_author_bond(Origin::signed(RootAccount::get()), 10));
+		assert_ok!(SimpleStaking::set_allowed_author_count(Origin::signed(RootAccount::get()), 1));
+		assert_ok!(SimpleStaking::register_as_author(Origin::signed(1)));
+
+		let addition = AuthorInfo {
+			who: 1,
+			deposit: 10,
+			last_block: None
+		};
+		assert_eq!(Balances::free_balance(1), 90);
+		assert_noop!(
+			SimpleStaking::register_as_author(Origin::signed(1)),
+			Error::<Test>::TooManyAuthors
+		);
+
+		assert_ok!(SimpleStaking::set_allowed_author_count(Origin::signed(RootAccount::get()), 5));
+		assert_noop!(
+			SimpleStaking::register_as_author(Origin::signed(1)),
+			Error::<Test>::AlreadyAuthor
+		);
+		assert_noop!(
+			SimpleStaking::register_as_author(Origin::signed(RootAccount::get())),
+			BalancesError::<Test>::InsufficientBalance
+		);
+		assert_eq!(SimpleStaking::authors(), vec![addition]);
```

Comment:
```suggestion
		assert_ok!(SimpleStaking::set_author_bond(Origin::signed(RootAccount::get()), 10));
		assert_ok!(SimpleStaking::set_allowed_author_count(Origin::signed(RootAccount::get()), 1));
		
		// can add 1 new 
		assert_ok!(SimpleStaking::register_as_author(Origin::signed(1)));
		let addition = AuthorInfo {
			who: 1,
			deposit: 10,
			last_block: None
		};
		assert_eq!(SimpleStaking::authors(), vec![addition]);
		assert_eq!(Balances::free_balance(1), 90);
		
		// but no more
		assert_noop!(
			SimpleStaking::register_as_author(Origin::signed(1)),
			Error::<Test>::TooManyAuthors
		);

		// increase limit
		assert_ok!(SimpleStaking::set_allowed_author_count(Origin::signed(RootAccount::get()), 5));
		// but still won't accept dupe..
		assert_noop!(
			SimpleStaking::register_as_author(Origin::signed(1)),
			Error::<Test>::AlreadyAuthor
		);
		// or poor author (3 is not endowed)
		assert_noop!(
			SimpleStaking::register_as_author(Origin::signed(3)),
			BalancesError::<Test>::InsufficientBalance
		);
		// but an endowed account works now
		assert_ok!(SimpleStaking::register_as_author(Origin::signed(2)));
		
		assert_eq!(SimpleStaking::authors().len(), 2);
```

Much better, do you agree? each test should clearly revolve around a particular scenario and cover its success and fail cases in an understandable manner. I think the previous version covered most cases but: 

1. it was a bit intertwined: you added ` 1` as author at the beginning, then check some other stuff, then assert `assert_eq!(SimpleStaking::authors(), vec![addition]);` at the end. I think this assertion belongs to the phase where we add 1. 
2. Some docs are good to show your trait of thoughts.
3. I think adding a second successful author would have been good which I added.

Reply:
ya, I was trying to be clever and just wanted to check that anything else done didn't effect the vector, but it just made the test a little disjointed.